### PR TITLE
CARD-compliance metadata and RFI information

### PR DESCRIPTION
**Pull request by seongsujeong**
This PR is to populate the  CARD-compliance metadata.

### CODE REVIEW

**Comment by seongsujeong** [comment]
@gshiroma I think I've populated as much as I can at the moment. I've left `TODO` comments where I want to revisit later. Do you want to take a first look?

File: src/rtc/h5_prep.py [hunk 1]
```diff
@@ -300,20 +309,195 @@ def get_metadata_dict(product_id: str,
...
+        'RTC/metadata/fileFormatSpecification/localIncidenceAngle/bitsPerSample':
+            ['local_inc_bits_per_sample', 32,
+             'Bytes per pixel'],
+        'RTC/metadata/fileFormatSpecification/localIncidenceAngle/byteOrder':
+            ['local_inc_byte_order', sys.byteorder,
+             'Byte order of the data'],
+
+        # file format spec. for backscatter - for 3.1
+        'RTC/metadata/fileFormatSpecification/backScatter/measurementType': # 3.1
+            ['measurement_type', 'Gamma-nought', 'Sample type of the data'],
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:448]:
```suggestion
            ['measurement_type', 'Gamma0', 'Sample type of the data'],
```

File: src/rtc/h5_prep.py [hunk 2]
```diff
@@ -300,20 +309,195 @@ def get_metadata_dict(product_id: str,
...
+            ['data_access',
+             'https://search.asf.alaska.edu/',
+             'Where the source data can be retrieved'],
+        'RTC/metadata/sourceDataInformation/radarBand': # 1.6.4
+            ['radar_band', 'C', 'Radar band'],
+        'RTC/metadata/sourceDataInformation/processingFacility': # 1.6.6
+            ['source_processing_facility',
+             'Sentinel-1 Instrument Processing Facility (IPF)',
+             'Source data processing facility'],
+        'RTC/metadata/sourceDataInformation/processingDate': # 1.6.6
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:339]:
```suggestion
        'RTC/metadata/sourceDataInformation/processingDateTime': # 1.6.6
```

File: src/rtc/h5_prep.py [hunk 3]
```diff
@@ -300,20 +309,195 @@ def get_metadata_dict(product_id: str,
...
+            ['source_processing_date',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:340]:
```suggestion
            ['source_processing_date_time',
```

File: src/rtc/h5_prep.py [hunk 4]
```diff
@@ -300,20 +309,195 @@ def get_metadata_dict(product_id: str,
...
+             burst_in.burst_misc_metadata.processing_date.strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
+             'Processing date of the source data'],
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:342]:
```suggestion
             'Processing date and time of the source data'],
```

**Review by gshiroma — COMMENTED** [review]
Thank you so much, @seongsujeong . This is really helpful. We talked offline, but I'll leave some of the comments we talked about here for reference.

File: src/rtc/h5_prep.py [hunk 5]
```diff
@@ -265,32 +268,38 @@ def get_metadata_dict(product_id: str,
         #    [mission_id, 'Mission identifier'],
         'identification/platform':
             ['platform', platform_id, 'Platform name'],
-        'identification/sensor':
-            ['sensor', platform_id, 'Sensor instrument name'],
-        'identification/productType':
+        # https://sentinel.esa.int/web/sentinel/missions/sentinel-1/satellite-description
+         'identification/sensor':
+            ['sensor', 'C-SAR', 'Sensor instrument name'],
```
[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/rtc/h5_prep.py:273]: The specification requires to mention the instrument to collect the data (section 1.6.2). The sensor name was came from the ESA link in the comment above.

File: src/rtc/h5_prep.py [hunk 6]
```diff
@@ -300,20 +309,195 @@ def get_metadata_dict(product_id: str,
...
+             burst_in.azimuth_time_interval,
+             'Azimuth spacing of the source data in seconds'], 
+        'RTC/metadata/sourceDataInformation/slantRangeSpacing': # 1.6.7
+            ['source_slant_range_spacing',
+             burst_in.range_pixel_spacing,
+             'Slant range spacing of the source data in meters'], 
+
+        'RTC/metadata/sourceDataInformation/azimuthResolution': # 1.6.7
+            ['source_azimuth_resolution',
+             burst_in.azimuth_time_interval,
```
Comment (seongsujeong) [inline src/rtc/h5_prep.py:376]: Please note that `*resolution` have the same values as `*spacing`

File: src/rtc/h5_prep.py [hunk 7]
```diff
@@ -300,20 +309,195 @@ def get_metadata_dict(product_id: str,
...
+
+        'RTC/metadata/processingInformation/noiseRemovalApplied': # 3.3
+            ['noise_removal_applied',
+             cfg_in.groups.processing.apply_thermal_noise_correction,
+             'A flag to indicate whether noise removal was applied'],
+
+        'RTC/metadata/processingInformation/demReference': # 4.2
+            ['dem_source_description', 'Copernicus DEM',
+             'DEM source description'], #TODO confirm, might need to be populated via runconfig
+
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:480]:
```suggestion
```

File: src/rtc/h5_prep.py [hunk 8]
```diff
@@ -371,11 +557,52 @@ def get_metadata_dict(product_id: str,
...
     }
+
+    # Add reference to the thermal noise correction algorithm when the correction is applied
+    if cfg_in.groups.processing.apply_thermal_noise_correction:  # 3.3
+        metadata_dict['RTC/metadata/processingInformation/noiseRemovalAlgorithmReference'] =\
+        ['noise_removal_algorithm_reference',
+         ('https://sentinels.copernicus.eu/documents/247904/2142675/'
+          'Thermal-Denoising-of-Products-Generated-by-Sentinel-1-IPF.pdf/'
+          '11d3bd86-5d6a-4e07-b8bb-912c1093bf91?t=1511973926000'),
+          'A reference to the noise removal algorithm applied']
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:570]:
```suggestion
    if cfg_in.groups.processing.apply_thermal_noise_correction:  # 3.3
        noise_removal_algorithm_reference = 'https://sentinels.copernicus.eu/documents/247904/2142675/'
          'Thermal-Denoising-of-Products-Generated-by-Sentinel-1-IPF.pdf/'
          '11d3bd86-5d6a-4e07-b8bb-912c1093bf91?t=1511973926000'
    else:
        noise_removal_algorithm_reference = '(noise removal not applied)'
    metadata_dict['RTC/metadata/processingInformation/noiseRemovalAlgorithmReference'] =\
        ['noise_removal_algorithm_reference',
         noise_removal_algorithm_reference,
          'A reference to the noise removal algorithm applied']
```

File: src/rtc/h5_prep.py [hunk 9]
```diff
@@ -371,11 +557,52 @@ def get_metadata_dict(product_id: str,
...
+        if cfg_in.groups.processing.rtc.algorithm_type == 'area_projection':
+            url_rtc_algorithm_document = 'https://ieeexplore.ieee.org/document/9695438'
+        elif cfg_in.groups.processing.rtc.algorithm_type == 'bilinear_distribution':
+            url_rtc_algorithm_document = 'https://ieeexplore.ieee.org/document/5752845'
+        else:
+            raise NotImplementedError
+        metadata_dict['RTC/metadata/processingInformation/rtcAlgorithmReference'] =\
+                ['rtc_algorithm_reference',
+                 url_rtc_algorithm_document,
+                'A reference to the RTC algorithm applied']
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:583]: Same idea here. Add something like "RTC not applied" if `cfg_in.groups.processing.apply_rtc` is `False`

File: src/rtc/h5_prep.py [hunk 10]
```diff
@@ -300,20 +309,195 @@ def get_metadata_dict(product_id: str,
...
+        'RTC/metadata/sourceDataInformation/dataAccess':
+            ['data_access',
+             'https://search.asf.alaska.edu/',
+             'Where the source data can be retrieved'],
+        'RTC/metadata/sourceDataInformation/radarBand': # 1.6.4
+            ['radar_band', 'C', 'Radar band'],
+        'RTC/metadata/sourceDataInformation/processingFacility': # 1.6.6
+            ['source_processing_facility',
+             'Sentinel-1 Instrument Processing Facility (IPF)',
+             'Source data processing facility'],
```
[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/rtc/h5_prep.py:338]: I've made the change on this PR and s1-reader. The value will look like for example:
>organization: "ESA", site: "Production Service-SERCO", country: "Italy"

File: src/rtc/h5_prep.py [hunk 11]
```diff
@@ -300,20 +309,195 @@ def get_metadata_dict(product_id: str,
...
+             'Slant range spacing of the source data in meters'], 
+
+        'RTC/metadata/sourceDataInformation/azimuthResolution': # 1.6.7
+            ['source_azimuth_resolution',
+             burst_in.azimuth_time_interval,
+             'Azimuth time resolution of the source data in seconds'],
+        'RTC/metadata/sourceDataInformation/slantRangeResolution': # 1.6.7
+            ['source_slant_range_resolution',
+             burst_in.range_pixel_spacing,
+             'Slant range resolution of the source data in meters'],
```
[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/rtc/h5_prep.py:381]: Took a look at the product annotation, and it seems the closest information is `processorScalingFactor` Below is the description from the product specification:
>Processor scaling factor. This value includes the input scaling factor and the range and azimuth oversampling factors from the auxiliary input and is applied multiplicatively to the image data during processing. This is the value referred to as kproc in [R-14]

File: src/rtc/h5_prep.py [hunk 12]
```diff
@@ -576,3 +802,109 @@ def save_hdf5_dataset(ds_filename, h5py_obj, root_path,
...
+
+    Parameters
+    ----------
+    h5py_obj: HDF5 object
+    burst: Sentinel1BurstSlc
+    rfi_root_path: str
+
+    '''
+    is_rfi_info_empty = burst.burst_rfi_info is None
+    dset = h5py_obj.create_dataset(f'{rfi_root_path}/isRfiInfoAvailable',
```
[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/rtc/h5_prep.py:817]: I think we can benefit from following the ESA's way of using acronym in the field name, because we can extract the metadata from the annotation file easily.
If you don't have strong opinion about capitalizing `RFI`, then I would like to follow the ESA's convention.

[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/rtc/h5_prep.py]: Another thought: I think we can use `RFI` in the fields that we've added, and use `'Rfi` if we extract the values from the annotation.

File: src/rtc/h5_prep.py (hunk 6 above)
Comment (gshiroma) [reply src/rtc/h5_prep.py:378]: If we don't have resolution values, let's comment these lines for now (we should populate them later with correct values).

File: src/rtc/h5_prep.py [hunk 13]
```diff
@@ -300,20 +309,196 @@ def get_metadata_dict(product_id: str,
...
+        #  sentinel-1-sar/acquisition-modes/interferometric-wide-swath)
+        'RTC/metadata/sourceDataInformation/maxNoiseEquivalentSigmaZero':  # 1.6.9
+            ['max_noise_equivalent_sigma_zero',
+             -22,
+             'Maximum Noise equivalent sigma0 in dB'], #TODO see if it can be extracted from burstNoise
+
+        'RTC/metadata/processingInformation/dataAccess':  # placeholder for 1.7.1
+            ['product_data_access',
+             'TBD',
+             'URL to access the product data'],
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:404]:
```suggestion
        # TODO Review: should we expose this parameter in the runconfig?
        #'RTC/metadata/processingInformation/dataAccess':  # placeholder for 1.7.1
        #    ['product_data_access',
        #     'TBD',
        #     'URL to access the product data'],
```
Let's comment this for now

File: src/rtc/h5_prep.py [hunk 14]
```diff
@@ -300,20 +309,196 @@ def get_metadata_dict(product_id: str,
...
+             'data format of the measurement'],
+        'RTC/metadata/fileFormatSpecification/backScatter/dataType':
+            ['measurement_data_type', 'Float',
+             'data type of the measurement'],
+        'RTC/metadata/fileFormatSpecification/backScatter/bitsPerSample':
+            ['measurement_bits_per_sample', 32,
+             'Bytes per sample in the measurement'],
+        'RTC/metadata/fileFormatSpecification/backScatter/byteOrder':
+            ['measurement_byte_order', sys.byteorder,
+             'Byte order of the measurement'],
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:469]:
```suggestion
```
Thank you, @seongsujeong .  I think we can remove these lines.  These layers (RTC-S1 imagery and radar geometry layers) will be provided as Cloud-Optimized GeoTIFFs. All these parameters should be part of the GeoTIFF specification with the exception of the "sample type"/"measurement type", which will be saved in the band description.

File: src/rtc/h5_prep.py [hunk 15]
```diff
@@ -300,20 +309,196 @@ def get_metadata_dict(product_id: str,
...
+
+        # The field below to be populated only when using power in dB scale
+        #'RTC/metadata/fileFormatSpecification/scalingConversion': # 3.2
+        #    ['',
+        #     ('equation to convert from pixel linear amplitude / power to'
+        #      'logarithmic decibel scale')],
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:475]:
```suggestion
```

We can remove this

File: src/rtc/h5_prep.py [hunk 16]
```diff
@@ -300,20 +309,196 @@ def get_metadata_dict(product_id: str,
...
+
+        'RTC/metadata/processingInformation/noiseRemovalApplied':  # 3.3
+            ['noise_removal_applied',
+             cfg_in.groups.processing.apply_thermal_noise_correction,
+             'A flag to indicate whether noise removal was applied'],
+    
+        'RTC/metadata/processingInformation/geoidReference':  # for 4.2
+            ['geoid_source_description', 'EGM2008', 'Geoid source description'], #TODO confirm, might need to be populated via runconfig
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:483]:
```suggestion
    
        #'RTC/metadata/processingInformation/geoidReference':  # for 4.2
        #    ['geoid_source_description', 'EGM2008', 'Geoid source description'], #TODO confirm, might need to be populated via runconfig
```
Let's comment this for now

File: src/rtc/h5_prep.py [hunk 17]
```diff
@@ -300,20 +309,196 @@ def get_metadata_dict(product_id: str,
...
+            ['absolute_accuracy_northing',
+             [0.0, 0.0],
+             ('An estimate of the absolute localisation error in north direction'
+              'provided as bias and standard deviation')],
+
+        'RTC/grids/processingInformation/absoluteAccuracyEasting':  # placeholder for 4.3
+            ['absolute_accuracy_easting',
+             [0.0, 0.0],
+             ('An estimate of the absolute localisation error in east direction'
+              'provided as bias and standard deviation')],
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:495]: Let's comment these lines as well.

File: src/rtc/h5_prep.py [hunk 18]
```diff
@@ -371,11 +558,57 @@ def get_metadata_dict(product_id: str,
             ['config_files', cfg_in.run_config_path,
              'List of input config files used'],
         'RTC/metadata/processingInformation/inputs/demSource':
-            ['dem_source', dem_description, 'DEM source description']
+            ['dem_source', dem_description, 'DEM source Description'],
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:561]: We can revert this change.

File: src/rtc/h5_prep.py [hunk 19]
```diff
@@ -576,3 +816,124 @@ def save_hdf5_dataset(ds_filename, h5py_obj, root_path,
...
+    Parameters
+    ----------
+    burst_in: Sentinel1BurstSlc
+    rfi_root_path: str
+
+    '''
+    rfi_metadata_dict = {}
+
+    is_rfi_info_empty = burst_in.burst_rfi_info is None
+    rfi_metadata_dict[f'{rfi_root_path}/isRFIInfoAvailable'] =\
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:834]:
```suggestion
    rfi_metadata_dict[f'{rfi_root_path}/isRfiInfoAvailable'] =\
```
This follows the convention you are using below.

File: src/rtc/h5_prep.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (gshiroma) [reply src/rtc/h5_prep.py]: Sounds good, @seongsujeong , I'm fine with having `Rfi`. I didn't quite understand the second thought.

File: src/rtc/h5_prep.py [hunk 20]
```diff
@@ -265,6 +268,7 @@ def get_metadata_dict(product_id: str,
         #    [mission_id, 'Mission identifier'],
         'identification/platform':
             ['platform', platform_id, 'Platform name'],
+        # https://sentinel.esa.int/web/sentinel/missions/sentinel-1/satellite-description
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:271]: So it looks like the S-1 sensor is indeed called `CSAR`?

**Review by gshiroma — COMMENTED** [review]
This is really great, @seongsujeong . Thank you so much for working on that. i think we are almost there. I just did a few suggestions below that are probably be easy to address.

File: src/rtc/h5_prep.py (hunk 20 above)
Comment (seongsujeong) [reply src/rtc/h5_prep.py:271]: Yes, according to the ESA webpage. But I think it's okay to use `platform_id` here as we discussed.

File: src/rtc/h5_prep.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/rtc/h5_prep.py]: For example, the field name `isRFIInfoAvailable` above is not in the RFI annotation. I've added this field to explicitly indicate the existence of RFI information.
I thought we can use the capitalized acronym in that case.

**Comment by seongsujeong** [comment]
@gshiroma Thank you for the comment and suggestion to this PR. Please note that this PR is not only for CARD, but also adding RFI information into the product. I'll change the PR's title accordingly.

Also it would be great if you can take another look at the PR in `s1-reader` below. We need this this PR to populate RFI information into RTC product (and same for CSLC)
https://github.com/opera-adt/s1-reader/pull/97

**Review by gshiroma — COMMENTED** [review]
Thank you, @seongsujeong . I have a few more comments below.

File: src/rtc/h5_prep.py [hunk 21]
```diff
@@ -300,20 +309,137 @@ def get_metadata_dict(product_id: str,
...
+             (f'organization: \"{burst_in.burst_misc_metadata.processing_info_dict["organisation"]}\", '
+              f'site: \"{burst_in.burst_misc_metadata.processing_info_dict["site"]}\", '
+              f'country: \"{burst_in.burst_misc_metadata.processing_info_dict["country"]}\"'),
+             'Source data processing facility'],
+        'RTC/metadata/sourceDataInformation/processingDateTime':  # 1.6.6
+            ['source_processing_date_time',
+             burst_in.burst_misc_metadata.processing_info_dict['stop'],
+             'Processing date and time of the source data'],
+        'RTC/metadata/sourceDataInformation/processingSoftwareVersion':  # 1.6.6
+            ['source_processing_sw_version',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:347]:
```suggestion
        'RTC/metadata/sourceDataInformation/softwareVersion':  # 1.6.6
            ['source_data_software_version',
```

File: src/rtc/h5_prep.py [hunk 22]
```diff
@@ -300,20 +309,137 @@ def get_metadata_dict(product_id: str,
...
+        #'identification/CEOSDocumentIdentifier':
+        #    ["https://ceos.org/ard/files/PFS/NRB/v5.5/CARD4L-PFS_NRB_v5.5.pdf",
+        #     'Product version'],
+        'RTC/metadata/sourceDataInformation/numberOfAcquisitions': # 1.6.4
+            ['number_of_acquisitions',
+             1,
+             'Number of source data acquisitions'],
+        # TODO Review: should we expose this parameter in the runconfig?
+        'RTC/metadata/sourceDataInformation/dataAccess':
+            ['data_access',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:331]:
```suggestion
            ['source_data_access',
```

File: src/rtc/h5_prep.py [hunk 23]
```diff
@@ -300,20 +309,137 @@ def get_metadata_dict(product_id: str,
...
+             'https://search.asf.alaska.edu/',
+             'Data access URL'],
+        'RTC/metadata/sourceDataInformation/radarBand':  # 1.6.4
+            ['radar_band', 'C', 'Radar band'],
+        'RTC/metadata/sourceDataInformation/processingFacility': #  1.6.6
+            ['source_processing_facility',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:337]:
```suggestion
            ['source_data_processing_facility',
```

File: src/rtc/h5_prep.py [hunk 24]
```diff
@@ -300,20 +309,137 @@ def get_metadata_dict(product_id: str,
...
+             (f'organization: \"{burst_in.burst_misc_metadata.processing_info_dict["organisation"]}\", '
+              f'site: \"{burst_in.burst_misc_metadata.processing_info_dict["site"]}\", '
+              f'country: \"{burst_in.burst_misc_metadata.processing_info_dict["country"]}\"'),
+             'Source data processing facility'],
+        'RTC/metadata/sourceDataInformation/processingDateTime':  # 1.6.6
+            ['source_processing_date_time',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:343]:
```suggestion
            ['source_data_processing_date_time',
```

File: src/rtc/h5_prep.py [hunk 25]
```diff
@@ -300,20 +309,137 @@ def get_metadata_dict(product_id: str,
...
+             burst_in.burst_misc_metadata.processing_info_dict['stop'],
+             'Processing date and time of the source data'],
+        'RTC/metadata/sourceDataInformation/processingSoftwareVersion':  # 1.6.6
+            ['source_processing_sw_version',
+             str(burst_in.ipf_version),
+             'IPF version of the source data'],
+        
+        'RTC/metadata/sourceDataInformation/azimuthLooks':  # 1.6.6
+            ['source_azimuth_looks',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:352]:
```suggestion
            [None,
```
Let's remove this from the GeoTIFF metadata

File: src/rtc/h5_prep.py [hunk 26]
```diff
@@ -300,20 +309,137 @@ def get_metadata_dict(product_id: str,
...
+             burst_in.burst_misc_metadata.azimuth_looks,
+             'Azimuth number of looks'],
+        'RTC/metadata/sourceDataInformation/slantRangeLooks':  # 1.6.6
+            ['source_slant_range_looks',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:356]:
```suggestion
            [None,
```
Let's also remove this from the GeoTIFF metadata

File: src/rtc/h5_prep.py [hunk 27]
```diff
@@ -300,20 +309,137 @@ def get_metadata_dict(product_id: str,
...
+             burst_in.burst_misc_metadata.slant_range_looks,
+             'Slant range number of looks'],
+
+        'RTC/metadata/sourceDataInformation/productLevel':  # 1.6.6
+            ['source_product_level',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:361]:
```suggestion
            ['source_data_product_level',
```

File: src/rtc/h5_prep.py [hunk 28]
```diff
@@ -300,20 +309,137 @@ def get_metadata_dict(product_id: str,
...
+             'L1',
+             'Product level of the source data'],
+        'RTC/metadata/sourceDataInformation/geometry':  # 1.6.7
+            ['source_geometry',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:365]:
```suggestion
            ['source_data_geometry',
```

File: src/rtc/h5_prep.py [hunk 29]
```diff
@@ -300,20 +309,137 @@ def get_metadata_dict(product_id: str,
...
+             'slant range',
+             'Geometry of the source data'],
+        'RTC/metadata/sourceDataInformation/azimuthSpacing':  # 1.6.7
+            ['source_azimuth_spacing',
+             burst_in.azimuth_time_interval,
+             'Azimuth spacing of the source data in seconds'], 
+        'RTC/metadata/sourceDataInformation/slantRangeSpacing':  # 1.6.7
+            ['source_slant_range_spacing',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:373]:
```suggestion
            ['source_data_slant_range_spacing',
```

File: src/rtc/h5_prep.py [hunk 30]
```diff
@@ -300,20 +309,137 @@ def get_metadata_dict(product_id: str,
...
+        #    ['source_azimuth_resolution',
+        #     burst_in.azimuth_time_interval,
+        #     'Azimuth time resolution of the source data in seconds'],
+        'RTC/metadata/sourceDataInformation/slantRangeResolution':  # 1.6.7
+            ['source_slant_range_resolution',
+             burst_in.range_pixel_spacing,
+             'Slant range resolution of the source data in meters'],
+
+        'RTC/metadata/sourceDataInformation/nearRangeIncidenceAngle':  # 1.6.7
+            ['near_range_incidence_angle',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:387]:
```suggestion
            [None,
```

File: src/rtc/h5_prep.py [hunk 31]
```diff
@@ -300,20 +309,137 @@ def get_metadata_dict(product_id: str,
...
+             burst_in.burst_misc_metadata.inc_angle_near_range,
+             'Near range incidence angle in meters'],
+        'RTC/metadata/sourceDataInformation/farRangeIncidenceAngle':  # 1.6.7
+            ['far_range_incidence_angle',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:391]:
```suggestion
            [None,
```

File: src/rtc/h5_prep.py [hunk 32]
```diff
@@ -300,20 +309,137 @@ def get_metadata_dict(product_id: str,
...
+             burst_in.burst_misc_metadata.inc_angle_far_range,
+             'Far range incidence angle in meters'],
+        # Source for the max. NESZ:
+        # (https://sentinels.copernicus.eu/web/sentinel/user-guides/
+        #  sentinel-1-sar/acquisition-modes/interferometric-wide-swath)
+        'RTC/metadata/sourceDataInformation/maxNoiseEquivalentSigmaZero':  # 1.6.9
+            ['max_noise_equivalent_sigma_zero',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:398]:
```suggestion
            [None,
```

File: src/rtc/h5_prep.py [hunk 33]
```diff
@@ -300,20 +309,137 @@ def get_metadata_dict(product_id: str,
...
+        'RTC/metadata/processingInformation/parameters/FilteringApplied':  # 1.7.4
+            ['filtering_applied',
+            False,
+             'Flag if filter has been applied'],
+        'RTC/grids/imageDimensions':  # 1.7.7
+            ['product_image_dimensions',
+             np.array([cfg_in.geogrids[str(burst_in.burst_id)].length,
+                       cfg_in.geogrids[str(burst_in.burst_id)].width]),
+             'List indicating the number of lines and samples the RTC-S1 imagery and secondary layers'],
+
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:416]:
```suggestion
```
I think we don't need this.

File: src/rtc/h5_prep.py (hunk 20 above)
Comment (seongsujeong) [reply src/rtc/h5_prep.py:271]: Sounds good. I'll update the reference to the PDF document you posted above.

**Comment by seongsujeong** [comment]
A note on the test failure on this PR:
The error comes from `s1-reader`. In order to make this PR work properly, we need the PR #97 in `s1-reader` merged, and modify the test setup to pull the most recent version of the s1-reader. Currently the test if using 0.1.6.

File: src/rtc/h5_prep.py [hunk 34]
```diff
@@ -300,20 +310,126 @@ def get_metadata_dict(product_id: str,
...
+             (f'organization: \"{burst_in.burst_misc_metadata.processing_info_dict["organisation"]}\", '
+              f'site: \"{burst_in.burst_misc_metadata.processing_info_dict["site"]}\", '
+              f'country: \"{burst_in.burst_misc_metadata.processing_info_dict["country"]}\"'),
+             'Source data processing facility'],
+        'RTC/metadata/sourceData/processingDateTime':  # 1.6.6
+            ['source_data_processing_date_time',
+             burst_in.burst_misc_metadata.processing_info_dict['stop'],
+             'Processing date and time of the source data'],
+        'RTC/metadata/sourceData/processingSoftwareVersion':  # 1.6.6
+            ['source_processing_sw_version',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:348]:
```suggestion
        'RTC/metadata/sourceData/softwareVersion':  # 1.6.6
            ['source_data_software_version',
```

File: src/rtc/h5_prep.py [hunk 35]
```diff
@@ -300,20 +310,126 @@ def get_metadata_dict(product_id: str,
...
+        #     'Product version'],
+        'RTC/metadata/sourceData/numberOfAcquisitions': # 1.6.4
+            ['number_of_acquisitions',
+             1,
+             'Number of source data acquisitions'],
+        # TODO Review: should we expose this parameter in the runconfig?
+        'RTC/metadata/sourceData/dataAccess':
+            ['source_data_access',
+             'https://search.asf.alaska.edu/',
+             'Data access URL'],
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:334]:
```suggestion
        # TODO Review: should we expose this parameter in the runconfig?
        # 'RTC/metadata/sourceData/dataAccess':
        #     ['source_data_access',
        #      'https://search.asf.alaska.edu/',
        #      'Data access URL'],
```

**Review by gshiroma — COMMENTED** [review]
Thank you, @seongsujeong . I have two minor comments below but I think this PR is ready for merging. I'll just wait for your s1-reader PR to be merged so we can safely approve and merge this PR as well.

File: src/rtc/h5_prep.py (hunk 21 above)
Comment (seongsujeong) [reply src/rtc/h5_prep.py]: Skipping this comment as the comment below [here](https://github.com/opera-adt/RTC/pull/34#discussion_r1171861437) is more recent than this one.

File: src/rtc/h5_prep.py [hunk 36]
```diff
@@ -302,18 +312,124 @@ def get_metadata_dict(product_id: str,
...
         'identification/softwareVersion':
             ['software_version', str(SOFTWARE_VERSION), 'Software version'],
-        # 'identification/CEOSDocumentIdentifier':
-        #     ["https://ceos.org/ard/files/PFS/NRB/v5.5/CARD4L-PFS_NRB_v5.5.pdf",
-        #      'CEOS document identifier'],
+        #'identification/CEOSDocumentIdentifier':
+        #    ["https://ceos.org/ard/files/PFS/NRB/v5.5/CARD4L-PFS_NRB_v5.5.pdf",
+        #     'Product version'],
+        'RTC/metadata/sourceData/numberOfAcquisitions': # 1.6.4
+            ['number_of_acquisitions',
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:327]:
```suggestion
            ['source_data_number_of_acquisitions',
```

File: src/rtc/h5_prep.py [hunk 37]
```diff
@@ -302,18 +312,124 @@ def get_metadata_dict(product_id: str,
...
+             'Slant range number of looks'],
+
+        'RTC/metadata/sourceData/productLevel':  # 1.6.6
+            ['source_data_product_level',
+             'L1',
+             'Product level of the source data'],
+        'RTC/metadata/sourceData/geometry':  # 1.6.7
+            ['source_data_geometry',
+             'slant range',
+             'Geometry of the source data'],
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:368]:
```suggestion
        # 'RTC/metadata/sourceData/geometry':  # 1.6.7
        #     ['source_data_geometry',
        #     'slant range',
        #     'Geometry of the source data'],
```
Let's comment this out

File: src/rtc/h5_prep.py [hunk 38]
```diff
@@ -302,18 +312,124 @@ def get_metadata_dict(product_id: str,
...
+        #'RTC/metadata/processingInformation/dataAccess':  # placeholder for 1.7.1
+        #    ['product_data_access',
+        #     'TBD',
+        #     'URL to access the product data'],
+        'RTC/metadata/processingInformation/parameters/postProcessingFilteringApplied':  # 1.7.4
+            ['post_processing_filtering_applied',
+            False,
+             'Flag to indicate if post-processing filtering has been applied'],
+
+        'RTC/metadata/processingInformation/noiseCorrectionApplied':  # 3.3
```
Comment (gshiroma) [inline src/rtc/h5_prep.py:413]:
```suggestion
        'RTC/metadata/processingInformation/parameters/noiseCorrectionApplied':  # 3.3
```

**Review by gshiroma — COMMENTED** [review]
A few more comments.

**Review by gshiroma — COMMENTED** [review]
A few more comments.

**Review by gshiroma — APPROVED** [review]
LGTM. Thank you, @seongsujeong .
